### ISSUE TITLE
Box drawing tool

### ISSUE THREAD

**Issue by 0ion9** [mypaint/mypaint#167]
A box / rectangle drawing tool would be pretty handy for establishing proportions and 3d form.
I have the core of one here, but am not that settled about how interaction should work.
This version can be pasted into gui/linemode.py after dynamic_straight_line(), and for the moment simply patches itself over dynamic_line (meaning that shift-dragging in normal mode draws boxes, line mode draws boxes, and connected-line mode draws connected boxes)

```
    # draw boxes, optionally in perspective.
    def dynamic_box(self, x, y, sx, sy, radius_mul=1.0, intersection_offset=0.0, right_angles=False):
        corners = [(x,y), None, (sx, sy), None]
        length, nx, ny = length_and_normal(x, y, sx, sy)
        cx, cy = midpoint(x, y, sx, sy)

        entry_p, midpoint_p, junk, prange2, head, tail = self.line_settings()
        if right_angles:
            corners[1] = (x, sy)
            corners[3] = (sx, y)
        else:
            radius = (length / 2.0) * radius_mul
            offset = min(0.9999, max(0.0001, 0.5 + intersection_offset))
            cx, cy = multiply_add(x, y, nx, ny, offset * length)
            px, py = perpendicular(nx,ny)
            corners[1] = (cx + (px*radius), cy + (py*radius))
            corners[3] = (cx + (px*-radius), cy + (py*-radius))
        # if you leave brush_prep() call out, you get a sort of graph-drawing interface, 
        # where consecutive boxes are overlaid on each other. Is it possible to usably expose that?
        #ensure clockwiseness
        corners.sort(key=lambda k:get_angle(cx, cy, k[0], k[1]))
        self.brush_prep(corners[0][0], corners[0][1])
        for xx, yy in corners[1:]:#(corners[1], corners[3], corners[2], corners[0]):
            self._stroke_to(xx, yy, midpoint_p)
        self._stroke_to(corners[0][0], corners[0][1], midpoint_p)
        # XXX for some reason, the brush fades out on the final stroke, although I asked for midpoint_p on all strokes.

    dynamic_straight_line = dynamic_box
```

As noted in the comment, there is a bug where the brush fades out on the final stroke that I haven't tracked down yet, but this is a usable proof of concept. 
The interaction is currently based on the idea of drawing an imaginary line that defines two of the box corners. The other two are determined using the classic X technique (drawing an imaginary line that intersects the center of the first imaginary line)
If  right_angles is True, a much simpler box is drawn (axis-aligned bounding box)

small gallery of example output is [here](http://imgur.com/d5fAwfn,M3nA4ro); the first image was drawn subject to a bug that I had fixed by the time I did the second image.

This is very experimental at the moment. 

* I'm looking at the possibility of a two-stage interaction where you drag out the major axis and then the minor axis
(think of this in terms of describing an L,then flipping that L to produce the opposite two edges of the box.)
* the option for parallelogram drawing might have to go away
* I'd like to hook it into the perspective tool if that gets into master.. it would vastly improve mypaint's ability to quickly describe spaces.
* aforementioned bug where the final stroke fades out despite the code specifying constant pressure.

Feedback would be appreciated.

**Comment by 0ion9** [comment]
Don't really want to scope-creep, but i'm getting the increasing idea that parametrizable shapes are really useful. eg. [Here's](http://i.imgur.com/0TN0nZV.png) an example of how fast you can throw things together with armatures (code [here](https://bpaste.net/raw/ef633dfae57f)). 

I think it would actually be relatively easy to make the existing framework accept arbitrary functions for drawing. Interaction with shape parameters requires more thought, though -- it seems reasonable to fix the maximum number of shape parameters at two plus the underlying line(s), as this is the maximum you can adjust with a simple shift+drag (up/down | left/right). It would be good but obviously not necessary to also support some level of 'static' option (eg. different variations on armatures are variously good for kinematics, measuring, and establishing a space.)

**Comment by achadwick** [comment]
Firstly, no time for this before the 1.2 release.

Second, if we do it, it would need to be put into its own Box Tool module, not cobbled into `gui.linemode`. The existing `linemode.py` code is a mess because it came along at the same time as the fancy new mode stack stuff, and had to be integrated - and I think I did quite a bad job of it. The line drawing and state management stuff is pretty complicated too, and I've been too scared to try refactoring it.

Really I'd want all kinds of line tool to go into their own module, or at least their own tool class. The factoring isn't right in there at present.

You're right that parametric shapes are well outside the scope of a simple box tool. Given that all those vectors would have to be converted into pixels at some point you'd need some sort of draft→tweak→commit workflow like we'd need for #13. I guess.

I'm deferring this on master for the sake of planning the 1.2 release. Meanwhile, hopefully the code already in there for drawing pretty, interactive lines and buttons on the canvas will mature nicely for Frame and Symmetry (and #13), so improved line or new box tools should find it easy to benefit from those.

**Comment by 0ion9** [comment]
Yeah, I recognize that the design is kind of flimsy at the moment. Didn't mean to suggest this should have anything to do with 1.2 -- I wasn't even aware that 1.2 was planned to happen soon.

A more modest function -- and still very useful -- might be to provide a stroke function on SVG layers,  that creates a new layer and traces all strokes found in the SVG layer with the current brush, onto that new layer. This would go some way to addressing the fact that SVG imports are not easy to just 'fit in', since they are not rendered with mypaint brush engine. I'll look into that, in the mean time, and open an issue when i get somewhere usable with it.

**Comment by jamshark70** [comment]
FWIW, I find it quite amusing that MyPaint goes to the trouble of providing an oval-drawing tool (which is harder than drawing a rectangle), but AFAICS there is no simple rectangle.

The oval tool already does brush selection etc. so you're 99% of the way there, except for straight lines.

From this user's perspective, it's a baffling design decision to omit rectangles.

**Comment by odysseywestra** [comment]
@0ion9 Could you move this thread over to https://community.mypaint.org/c/feature-requests? I'm working on cleaning up the issue tracker by move all feature requests there so we can get community feedback. If you can also set your post as a wiki that way we can edit the scope of the feature request that would be awesome as well. Don't forget to link back to here as well so I can keep track of it incase we need to reopen the issue. Thanks.

**Comment by einarpersson** [comment]
I'm not keen on signing up for another community forum, but as a new user trying MyPaint out I just want to express that this struck me as weird and and a very unexpected and unfortunate limitation.
